Add jumptoanywhere command for easymotion

jsonMartin opened the pull request:
<!--
Yay! Thanks for sending us a PR! 🎊

Please ensure your PR adheres to:

- [X] Commit messages has a short & issue references when necessary
- [X] Each commit does a logical chunk of work.
- [X] It builds and tests pass (e.g `gulp`)
-->

**What this PR does / why we need it**:
This PR adds the missing `jumptoanywhere` command to the `easymotion` plugin. `jumptoanywhere` allows quick text navigation, and is basically a combination of `bd-w` and `bd-j` with some added behavior.

According to the easymotion VIM docs, the default behavior for this command is: `Beginning and End of word, camelCase, after '_', and after '#'`. The current implementation matches this expected behavior.

Here's a demo of what it looks like in action (_note my space bar is mapped to <Leader><Leader><Leader>j_):
![2018-03-19 18 18 59](https://user-images.githubusercontent.com/4033621/37630524-4c67560c-2ba2-11e8-9e37-d507fb14a106.gif)

The user can also configure custom jumptoanywhere regex to fine tune the behavior. For example, this is my current setting in VSCode, which also adds the first/last char of the line and `//` comments to the normal behavior (note the double escaping necessary in the config file):
```
"vim.easymotionJumptoanywhereRegex": "^\\s*.|\\b[A-Za-z0-9]|[A-Za-z0-9]\\b|_.|\\#.|[a-z][A-Z]|//"
```
**Which issue(s) this PR fixes**
https://github.com/VSCodeVim/Vim/issues/2434#issuecomment-373471794
<!--
Commits in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)
-->

**Special notes for your reviewer**:
This PR has been updated to also include custom regex settings via user preferences to match the behavior of `Easymotion_re_anywhere` that can be set in .vimrc

jsonMartin commented:
Just added another commit allowing the addition of custom regex for jumptoanywhere in user preferences; this allows the jumptoanywhere integration to allow custom matching options a la the `Easymotion_re_anywhere` in `.vimrc` setting.

I'll update the PR description with the new information

jpoon commented on src/actions/plugins/easymotion/easymotion.cmd.ts:
```diff
@@ -299,8 +299,14 @@ export class EasyMotionWordMoveCommandBase extends BaseEasyMotionCommand {
-    return vimState.easyMotion.sortedSearch(position, new RegExp('\\w{1,}', 'g'), options);
+    const regex = this._options.jumptoanywhere
+      ? configuration.easymotionJumptoanywhereRegex
+        ? new RegExp(configuration.easymotionJumptoanywhereRegex, 'g')
```
Can you capitalize the `Anywhere`?

jpoon commented on src/actions/plugins/easymotion/easymotion.cmd.ts:
```diff
@@ -299,8 +299,14 @@ export class EasyMotionWordMoveCommandBase extends BaseEasyMotionCommand {
+      ? configuration.easymotionJumptoanywhereRegex
+        ? new RegExp(configuration.easymotionJumptoanywhereRegex, 'g')
+        : // Default behavior matches beginning & ending of word, camelCase, after _ and after #
+          new RegExp('\b[A-Za-z0-9]|[A-Za-z0-9]\b|_.|#.|[a-z][A-Z]', 'g')
```
Is this match intended:

For `MaybeThisWillDo`, the above regex will match the M, e, T, s, W, D

jpoon replied:
This is a better visualization of what I mean: https://regexr.com/3mo29

jpoon replied:
Ah, I see your comment about camelCase, this is probably expected behaviour, but just want to confirm.

jsonMartin replied:
Hey @jpoon - Yeah, this is definitely intended behavior to match CamelCase. Since the VIM version of `jumptoanywhere` always places the cursor at the end of the match, this behavior pairs perfectly with the `labelPosition: after` setting that was already present in the easymotion API 😎

Here's a sample of what that particular string looks like in action with the jumptoanywhere motion & current default regex:
![2018-03-23 11 43 46](https://user-images.githubusercontent.com/4033621/37848057-066fc3c8-2e90-11e8-82e8-14d2a4fbd8a5.gif)

jpoon commented on src/actions/plugins/easymotion/types.ts:
```diff
@@ -11,4 +12,5 @@ export interface EasyMotionCharMoveOpions extends EasyMotionMoveOptionsBase {
 
 export interface EasyMotionWordMoveOpions extends EasyMotionMoveOptionsBase {
   labelPosition?: LabelPosition;
+  jumptoanywhere?: jumptoanywhere;
```
Please camel case. Unless is it generally known as `jumptoanywhere` all lowercase?

jsonMartin replied:
So all lowercase `jumptoanywhere` is how the original easymotion creators describe it in their vim documentation, so I was just trying to be consistent with the original implementation:
<img width="608" alt="screen shot 2018-03-23 at 12 09 10 pm" src="https://user-images.githubusercontent.com/4033621/37848987-3bd6b744-2e93-11e8-925a-947ad5f06840.png">

However, I agree that the camelcase naming convention is much easier to read and more JS-like. I have went ahead and updated all JS code references to be camelcased named, and only left the original `jumptoanywhere` text as documentation to describe the original matching motion found in standard VIM.

jsonMartin commented:
Ok @jpoon , I have implemented the requested changes with camelCase naming convention.

Also while preparing that commit, I fixed another bug I just discovered that arose when I refactored the default Regex from a Regex Literal to use `new Regex()` instead (forgot to double escape `\\b` when converting to a string). I might not have noticed it TBH unless you brought up regex, so thanks for that! Didn't originally notice this issue on my end because my custom jumpToAnywhere regex in preferences was working as intended! 😆

Tested it again, and it's working perfectly for me using the default behavior or custom regex. 

Thanks again for the feedback!

jpoon commented on README.md:
```diff
@@ -422,6 +422,7 @@ Motion Command | Description
 `<leader><leader><leader> bdw`|Start of word
 `<leader><leader><leader> bde`|End of word
 `<leader><leader><leader> bdjk`|Start of line
+`<leader><leader><leader> j`|Jumptoanywhere motion; default behavior matches beginning & ending of word, camelCase, after _ and after #
```
`Jumptoanywhere ` Casing :)

jpoon commented on src/configuration/configuration.ts:
```diff
@@ -203,6 +203,7 @@ class Configuration implements IConfiguration {
   easymotionMarkerFontWeight = 'normal';
   easymotionMarkerYOffset = 0;
   easymotionKeys = 'hklyuiopnm,qwertzxcvbasdgjf;';
+  easymotionJumpToAnywhereRegex = '';
```
Let's set the default here: `\\b[A-Za-z0-9]|[A-Za-z0-9]\\b|_.|#.|[a-z][A-Z]`.

jpoon replied:
You'll also need to update the `package.json` to add this value. Check out the comment at the top of this class.

jsonMartin replied:
@jpoon Great idea. Much cleaner logic in `easymotion.cmd.ts` now :)

jsonMartin commented:
Hey @jpoon, finally had a chance to catch up on things today, have been under the weather these past few days.

I have just finished implementing the requested changes and pushed them up 😎

xconverge commented:
Hope you are feeling better :) Thanks for getting back into it!

jpoon commented on src/actions/plugins/easymotion/types.ts:
```diff
@@ -1,4 +1,5 @@
 export type LabelPosition = 'after' | 'before';
+export type jumpToAnywhere = true | false;
```
Capital here please as we're exposing the type :)

jsonMartin replied:
OK, updated 👍
